ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -299,7 +299,7 @@ private class ContractsFetch(
 
     // skip if *we don't use the acs* (which can read past absEnd) and current
     // DB is already caught up to absEnd
-    if (startOffset == AbsoluteBookmark(absEnd.toDomain)) {
+    if (startOffset >= AbsoluteBookmark(absEnd.toDomain)) {
```

Comment:
Unfortunately, I think that this little optimization in #16401 might be getting you into more trouble than it's worth. The idea of `contractsFromOffsetIO` was that, if it returned, _at that time_ all the offsets were consistent with the returned bookmark. This little optimization already broke that assumption, so was probably unwise, by making it so only the _minimum_ is guaranteed to be the return value, and the change here widens the gap between intent and result even further.

Now, it would be a pain to deal with that by going all the way back to where `absEnd` is picked and picking again, but there is an appropriate place to reconsider: `fetchToAbsEnd` assumes all the response bookmarks for the individual templates are accurate, and if they don't line up, will rerun `fetchAndPersist` _specifically advancing `absEnd`_. While the original purpose of this was #8226/#8480, "advancing more than expected because of the ACS" is pretty similar to the scenario you're describing. The difference with your scenario is that `"it cannot go "too far" reading only the tx stream"` is no longer accurate; you need a new invariant of `contractsFromOffsetIO` to allow `fetchToAbsEnd` to go into the `feedback` branch of `fetchToAbsEnd` and rerun for that template at the later offset.